### PR TITLE
Poroelastic

### PR DESCRIPTION
**Pull request by sebwolf-de**
Finally, the poroelasticity PR ;-)
Major changes: 
* Space-Time-Predictor to integrate the stiff source term from the poroelastic equations
* Armadillo as an alternative to eigen3, as long as https://gitlab.com/libeigen/eigen/-/issues/2305 is not fixed

### CODE REVIEW

**Comment by uphoffc** [comment]
Why don't we use LAPACK? More likely to be installed than armadillo.

**Comment by sebwolf-de** [comment]
> Why don't we use LAPACK? More likely to be installed than armadillo.

Mostly, because armadillo is easier to use. After all, armadillo is just a fancy wrapper around LAPACK: https://scicomp.stackexchange.com/a/26397
The installation is really straightforward, so this does not pose a major issue IMHO. I hope that Eigen3 will fix the accuracy issue soonish, so that we can remove armadillo again.

**Comment by krenzland** [comment]
But LAPACK is available on many HPC systems as a module. This would make the installation easier.

**Comment by uphoffc** [comment]
When using LAPACK, one has to be a bit careful with the symbol names. Depending on the software stack one has to prepend N >= 0 underscores in front of the LAPACK function name. Fortunately CMake supports automatic name mangling in the following way:

```cmake
check_language(Fortran)
if(CMAKE_Fortran_COMPILER)
    enable_language(Fortran)
    include(FortranCInterface)
    FortranCInterface_HEADER(FC.h MACRO_NAMESPACE "FC_")

    find_package(LAPACK)
    if(${LAPACK_FOUND})
        add_definitions(-DHAVE_LAPACK)
    endif()
endif()
```

One can then create a little header file that mangles the name, e.g.:

```c++
#ifndef LAPACK_20210208_H
#define LAPACK_20210208_H

#include "FC.h"

#define FC_dsyev FC_GLOBAL(dsyev, DSYEV)
extern void FC_dsyev(char const* jobz, char const* uplo, int const* n, double* A, int const* lda,
                     double* W, double* work, int const* lwork, int* info);

#endif // LAPACK_20210208_H
```

**Review by krenzland — CHANGES_REQUESTED** [review]
Hi,
looks good mostly. Didn't spot any bugs.
Two things (in addition to Armadillo) are worth discussing:
- I'm not a fan of adding parameter files (and a test runner) in src/tests, especially because we already have a test runner in our CI: https://github.com/SeisSol/Examples/tree/master/convergence_elastic
- The Time kernel is nearly a direct copy of the elastic time kernel with some differences. This is really bad code quality. We can also tackle this together when I'll begin templating the kernels.

PS: Where are your matrices? Did you already push them to the git? Only asking because I always forget that step...

File: CMakeLists.txt [hunk 1]
```diff
@@ -263,6 +268,15 @@ if (MEMKIND)
   target_compile_definitions(SeisSol-lib PUBLIC USE_MEMKIND)
 endif()
 
+if(${EQUATIONS} STREQUAL "poroelastic")
+  find_package(Armadillo REQUIRED)
+  if(Armadillo_FOUND)
+    target_compile_definitions(SeisSol-lib PUBLIC HAS_ARMADILLO)
```
[the remark this answers is not in the document]
Comment (sebwolf-de) [reply CMakeLists.txt:274]: Armadillo is only required for poroelastic materials. But indeed, this is redundant with USE_POROELASTIC

File: azure-pipelines-seissol-template.yml [hunk 2]
```diff
@@ -36,7 +36,8 @@ jobs:
                             mechanisms=0
                         fi
                         mkdir -p build && cd build
-                        CMAKE_PREFIX_PATH=~ PKG_CONFIG_PATH=~/lib/pkgconfig/ cmake -DNETCDF=OFF -DMETIS=ON -DCOMMTHREAD=ON -DASAGI=OFF -DHDF5=ON -DCMAKE_BUILD_TYPE=${{ build_type }} -DTESTING=ON -DLOG_LEVEL=warning -DLOG_LEVEL_MASTER=info -DHOST_ARCH=hsw -DPRECISION=${{ precision }} -DEQUATIONS=${{ equation }} -DNUMBER_OF_MECHANISMS=$mechanisms -DGEMM_TOOLS_LIST=LIBXSMM ..
+                        CMAKE_PREFIX_PATH=~ PKG_CONFIG_PATH=~/lib/pkgconfig/ cmake -DNETCDF=OFF -DMETIS=ON -DORDER=6 -DCOMMTHREAD=ON -DASAGI=OFF -DHDF5=ON -DCMAKE_BUILD_TYPE=${{ build_type }} -DTESTING=ON -DLOG_LEVEL=warning -DLOG_LEVEL_MASTER=info -DHOST_ARCH=hsw -DPRECISION=${{ precision }} -DEQUATIONS=${{ equation }} -DNUMBER_OF_MECHANISMS=$mechanisms -DGEMM_TOOLS_LIST=LIBXSMM ..
                         make -j $(nproc)
+                        ./test_serial_test_suite
```
[the remark this answers is not in the document]
Comment (sebwolf-de) [reply azure-pipelines-seissol-template.yml:41]: True, I just put this in for debugging, as make test suppresses the output.

File: CMakeLists.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (krenzland) [reply CMakeLists.txt:274]: You can remove a lot of these ifdefs in the poor code - would make it easier to read

**Review by sebwolf-de — COMMENTED** [review]
Thanks for the comments, most of it should be addressed by now. I am still thinking about the LAPACK/armadillo thing...

**Review by krenzland — APPROVED** [review]

**Pull request merged**